application/src/main/java/org/thingsboard/server/service/install/DefaultDatabaseSchemaSettingsService.java

```diff
@@ -0,0 +1,146 @@
+        //TODO: remove after release (3.9.0)
+        createProductIfNotExists();
+
+        String dbSchemaVersion = getDbSchemaVersion();
+
+        if (DefaultDataUpdateService.getEnv("SKIP_SCHEMA_VERSION_CHECK", false)) {
+            log.info("Skipped DB schema version check due to SKIP_SCHEMA_VERSION_CHECK set to 'true'.");
+            return;
+        }
+
+        String product = getProductFromDb();
+        if (!CURRENT_PRODUCT.equals(product)) {
+            onSchemaSettingsError(String.format("Upgrade failed: can't upgrade ThingsBoard %s database using ThingsBoard %s.", product, CURRENT_PRODUCT));
+        }
+
+        if (dbSchemaVersion.equals(getPackageSchemaVersion())) {
+            onSchemaSettingsError("Upgrade failed: database already upgraded to current version. You can set SKIP_SCHEMA_VERSION_CHECK to 'true' if force re-upgrade needed.");
+        }
+
+        if (!SUPPORTED_VERSIONS_FOR_UPGRADE.contains(dbSchemaVersion)) {
+            onSchemaSettingsError(String.format("Upgrade failed: database version '%s' is not supported for upgrade. Supported versions are: %s.",
+                    dbSchemaVersion, SUPPORTED_VERSIONS_FOR_UPGRADE
+            ));
+        }
+    }
+
+    @Deprecated(forRemoval = true, since = "3.9.0")
+    private void createProductIfNotExists() {
+        boolean isCommunityEdition;
+        try {
+            jdbcTemplate.queryForObject("SELECT 1 FROM information_schema.tables WHERE table_name = 'integration'", Integer.class);
+            isCommunityEdition = false;
+        } catch (EmptyResultDataAccessException e) {
+            isCommunityEdition = true;
+        }
+        String product = isCommunityEdition ? "CE" : "PE";
+        jdbcTemplate.execute("ALTER TABLE tb_schema_settings ADD COLUMN IF NOT EXISTS product varchar(2) DEFAULT '" + product + "'");
+    }
+
+    @Override
+    public void createSchemaSettings() {
+        Long schemaVersion = getSchemaVersionFromDb();
+        if (schemaVersion == null) {
+            jdbcTemplate.execute("INSERT INTO tb_schema_settings (schema_version, product) VALUES (" + getPackageSchemaVersionForDb() + ", '" + CURRENT_PRODUCT + "')");
+        }
+    }
+
+    @Override
+    public void updateSchemaVersion() {
+        jdbcTemplate.execute("UPDATE tb_schema_settings SET schema_version = " + getPackageSchemaVersionForDb());
+    }
+
+    @Override
+    public String getPackageSchemaVersion() {
+        if (packageSchemaVersion == null) {
+            packageSchemaVersion = buildProperties.getVersion().replaceAll("[^\\d.]", "");
+        }
+        return packageSchemaVersion;
+    }
+
+    @Override
+    public String getDbSchemaVersion() {
+        if (schemaVersionFromDb == null) {
+            Long version = getSchemaVersionFromDb();
+            if (version == null) {
+                onSchemaSettingsError("Upgrade failed: the database schema version is missing.");
+            }
+
+            @SuppressWarnings("DataFlowIssue")
```

Comment:
What's this?)

Reply:
Suppress 
Unboxing of 'version' may produce 'NullPointerException'
but its impossible, because if `version` is null we invoke `onSchemaSettingsError`.